### PR TITLE
Async Agent

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
Use async API in the agent service

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
^^^

## Tracking Issue
_NA_

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit/extend/backend/agent_service.py [hunk 1]
```diff
@@ -17,37 +19,39 @@
...
+    async def CreateTask(self, request: CreateTaskRequest, context: grpc.ServicerContext) -> CreateTaskResponse:
         try:
             tmp = TaskTemplate.from_flyte_idl(request.template)
             inputs = LiteralMap.from_flyte_idl(request.inputs) if request.inputs else None
             agent = AgentRegistry.get_agent(context, tmp.type)
             if agent is None:
                 return CreateTaskResponse()
-            return agent.create(context=context, inputs=inputs, output_prefix=request.output_prefix, task_template=tmp)
+            return await asyncio.to_thread(
+                agent.create, context=context, inputs=inputs, output_prefix=request.output_prefix, task_template=tmp
```
Comment (kumare3) [inline flytekit/extend/backend/agent_service.py:31]: we should continue the async context downstream.

Maybe there is a better method, in the agent code
```python
class SnowflakeAgent:

def __init__():
   super().__init__(asynchronous=False)

def create():
   ...
```

And then here
async def CreateTask(...):
   if agent.asynchronous:
       return await agent.create(...)
   return asyncio.to_thread(agent.sync_create, ...)

File: flytekit/extend/backend/agent_service.py [hunk 2]
```diff
@@ -17,37 +19,47 @@
...
+    async def CreateTask(self, request: CreateTaskRequest, context: grpc.ServicerContext) -> CreateTaskResponse:
         try:
             tmp = TaskTemplate.from_flyte_idl(request.template)
             inputs = LiteralMap.from_flyte_idl(request.inputs) if request.inputs else None
             agent = AgentRegistry.get_agent(context, tmp.type)
             if agent is None:
                 return CreateTaskResponse()
-            return agent.create(context=context, inputs=inputs, output_prefix=request.output_prefix, task_template=tmp)
+            if agent.asynchronous:
+                return await agent.async_create(
```
Comment (kumare3) [inline flytekit/extend/backend/agent_service.py:31]: lets put a 
```python
log.debug()
try:
   agent.async
except Exception:
   log.error
   raise
log.debug(done)
```

Comment (kumare3) [reply flytekit/extend/backend/agent_service.py:31]: the idea is we want to log and know what went wrong in production. Think about how will you make it easier to debug in production and manage in production. Also we should think about adding prometheus metrics or open telemetry (separate PR is ok)

Comment (pingsutw) [reply flytekit/extend/backend/agent_service.py]: done

**Review by kumare3 — APPROVED** [review]

**Pull request merged**